### PR TITLE
mention `getindex()` shouldn't return a "view"

### LINKED ISSUE #47078: Can `getindex(A::AbstractArray, inds::AbstractArray)` result be a view?
It's not clear whether the `AbstractArray` interface allows implementations to define `getindex(A::AbstractArray, inds::AbstractArray)` so that it returns a view of `A` that is affected by later mutation of `A` or `inds`. This lead to a bug in DataFrames as [`filter` assumes that `inds` can be emptied](https://github.com/JuliaLang/julia/pull/31929/files#r987877494), which had the unintended effect of emptying the view (https://github.com/JuliaData/DataFrames.jl/pull/3192). This will be fixed soon in DataFrames anyway by copying `inds`, but it would be good to clarify what is expected by the interface and adjust either the manual or the `filter` implementation.

### PR DESCRIPTION
**Pull request by Moelf**
fix  #47078

### CODE REVIEW

**Comment by Moelf** [comment]
isn't the issue we agree it should never be a view but just didn't mention it?

**Comment by Moelf** [comment]
can we merge this?

**Comment by ViralBShah** [comment]
I'd like a few more eyes on it. @KristofferC Is this reasonable to merge?

**Comment by KristofferC** [comment]
Shouldn't this be in the documentation for `getindex` or something? It somehow doesn't feel like this is the right place for a side note like this.

**Comment by jishnub** [comment]
I'd prefer the phrasing to suggest what to do instead of warning against what not to do. A `view` is one of the many wrappers that `getindex` shouldn't return. It shouldn't return a reshaped array either, for example `getindex(a::AbstractArray, ::Colon) = reshape(a, Val(1))` isn't a `view` technically, but it shares the underlying memory with the parent array.

**Comment by Moelf** [comment]
@KristofferC this is for developers, for base vectors, `getindex` don't return view as outlined in multi dimensional array manual, and we already tell user to use `@view` if they don't want a copy.

File: doc/src/manual/interfaces.md [hunk 1]
```diff
@@ -152,10 +152,10 @@ julia> collect(Iterators.reverse(Squares(4)))
 
 | Methods to implement | Brief description                |
 |:-------------------- |:-------------------------------- |
-| `getindex(X, i)`     | `X[i]`, indexed element access   |
-| `setindex!(X, v, i)` | `X[i] = v`, indexed assignment   |
+| `getindex(X, i)`     | `X[i]`, indexed access, shouldn't return a [`view`](@ref)  |
```
Comment (nalimilan) [inline doc/src/manual/interfaces.md:155]:
```suggestion
| `getindex(X, i)`     | `X[i]`, indexed access, should allocate a copy  |
```
(Should also fix spacing.)

File: base/abstractarray.jl [hunk 2]
```diff
@@ -1256,6 +1256,9 @@ end
 
 Return a subset of array `A` as specified by `inds`, where each `ind` may be,
 for example, an `Int`, an [`AbstractRange`](@ref), or a [`Vector`](@ref).
+
+To index multiple elements without making a copy, use [`@view`](@ref).
```
Comment (nalimilan) [inline base/abstractarray.jl:1260]:
```suggestion
When `inds` selects multiple elements, this function returns a newly
allocated array. To index multiple elements without making a copy,
use [`view`](@ref) instead.
```

File: base/abstractarray.jl [hunk 3]
```diff
@@ -1256,6 +1256,9 @@ end
```
Comment (nalimilan) [inline base/abstractarray.jl:1256]: How about specifying `A::AbstractArray` on the line above while we're at it? The description below isn't guaranteed to match non-array objects.

**Comment by nalimilan** [comment]
Can we revive this? I hope my suggestions can help addressing the comments.

File: base/abstractarray.jl (hunk 3 above)
Comment (Moelf) [reply base/abstractarray.jl:1256]: that page is specific to array interface, i think thats implied

**Comment by adienes** [comment]
+1

File: doc/src/manual/interfaces.md [hunk 4]
```diff
@@ -152,10 +152,10 @@ julia> collect(Iterators.reverse(Squares(4)))
 
 | Methods to implement | Brief description                |
 |:-------------------- |:-------------------------------- |
-| `getindex(X, i)`     | `X[i]`, indexed element access   |
-| `setindex!(X, v, i)` | `X[i] = v`, indexed assignment   |
+| `getindex(X, i)`     | `X[i]`, indexed access, should allocate a copy  |
```
Comment (timholy) [inline doc/src/manual/interfaces.md:155]:
```suggestion
| `getindex(X, i)`     | `X[i]`, indexed access, non-scalar `i` should allocate a copy  |
```

**Comment by timholy** [comment]
Thanks!

**Pull request merged**